File: app/src/main/java/io/github/summerdez/asmrplayer/data/files/DocumentFiles.java
package io.github.summerdez.asmrplayer.data.files;

import io.github.summerdez.asmrplayer.R;
import io.github.summerdez.asmrplayer.data.*;
import io.github.summerdez.asmrplayer.data.remote.*;
import io.github.summerdez.asmrplayer.data.download.*;
import io.github.summerdez.asmrplayer.data.files.*;
import io.github.summerdez.asmrplayer.domain.*;
import io.github.summerdez.asmrplayer.domain.model.*;
import io.github.summerdez.asmrplayer.playback.*;
import io.github.summerdez.asmrplayer.presentation.*;
import io.github.summerdez.asmrplayer.ui.*;
import io.github.summerdez.asmrplayer.ui.activity.*;
import io.github.summerdez.asmrplayer.ui.components.*;
import io.github.summerdez.asmrplayer.ui.screens.*;
import io.github.summerdez.asmrplayer.ui.theme.*;
import io.github.summerdez.asmrplayer.ui.util.*;
import io.github.summerdez.asmrplayer.di.*;
import android.content.ClipData;
import android.content.ContentResolver;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.media.MediaMetadataRetriever;
import android.net.Uri;
import android.provider.DocumentsContract;
import android.provider.OpenableColumns;
import android.text.TextUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.io.IOException;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public final class DocumentFiles {
    private static final String[] FOLDER_COLUMNS = new String[]{
            DocumentsContract.Document.COLUMN_DOCUMENT_ID,
            DocumentsContract.Document.COLUMN_DISPLAY_NAME,
            DocumentsContract.Document.COLUMN_MIME_TYPE
    };

    private DocumentFiles() {
    }

    public static Intent audioPickerIntent(boolean allowMultiple) {
        Intent intent = baseOpenDocumentIntent("audio/*");
        if (allowMultiple) {
            intent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true);
        }
        return intent;
    }

    public static Intent folderPickerIntent() {
        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION
                | Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION
                | Intent.FLAG_GRANT_PREFIX_URI_PERMISSION);
        return intent;
    }

    public static Intent subtitlePickerIntent() {
        Intent intent = baseOpenDocumentIntent("*/*");
        intent.putExtra(Intent.EXTRA_MIME_TYPES, new String[]{
                "application/x-subrip",
                "text/plain",
                "text/*",
                "application/octet-stream"
        });
        return intent;
    }

    public static Intent imagePickerIntent() {
        return baseOpenDocumentIntent("image/*");
    }

    public static List<Uri> urisFromResult(Intent data) {
        List<Uri> uris = new ArrayList<>();
        ClipData clipData = data.getClipData();
        if (clipData != null) {
            for (int i = 0; i < clipData.getItemCount(); i++) {
                Uri uri = clipData.getItemAt(i).getUri();
                if (uri != null) {
                    uris.add(uri);
                }
            }
        } else if (data.getData() != null) {
            uris.add(data.getData());
        }
        return uris;
    }

    public static void persistReadPermission(Context context, Uri uri) {
        try {
            context.getContentResolver().takePersistableUriPermission(
                    uri,
                    Intent.FLAG_GRANT_READ_URI_PERMISSION);
        } catch (SecurityException ignored) {
        }
    }

    public static void persistTreeReadPermission(Context context, Intent data, Uri uri) {
        int flags = data == null ? Intent.FLAG_GRANT_READ_URI_PERMISSION : data.getFlags();
        int readFlags = flags & Intent.FLAG_GRANT_READ_URI_PERMISSION;
        if (readFlags == 0) {
            readFlags = Intent.FLAG_GRANT_READ_URI_PERMISSION;
        }
        try {
            context.getContentResolver().takePersistableUriPermission(uri, readFlags);
        } catch (SecurityException ignored) {
        }
    }

    public static List<FolderImportItem> folderAudioImports(Context context, Uri treeUri) {
        if (treeUri == null) {
            return Collections.emptyList();
        }
        List<TreeDocument> documents = treeDocuments(context, treeUri);
        Map<String, TreeDocument> subtitlesByName = new HashMap<>();
        for (TreeDocument document : documents) {
            if (isSupportedSubtitleName(document.name)) {
                subtitlesByName.put(normalizedName(document.name), document);
            }
        }

        List<FolderImportItem> imports = new ArrayList<>();
        for (TreeDocument document : documents) {
            if (!isSupportedAudioName(document.name, document.mimeType)) {
                continue;
            }
            TreeDocument subtitle = firstMatchingSubtitle(subtitlesByName, document.name);
            imports.add(new FolderImportItem(
                    document.name,
                    document.uri,
                    subtitle == null ? "" : subtitle.name,
                    subtitle == null ? null : subtitle.uri));
        }
        Collections.sort(imports, (left, right) -> left.audioName.compareToIgnoreCase(right.audioName));
        return imports;
    }

    public static String displayName(Context context, Uri uri) {
        if (ContentResolver.SCHEME_CONTENT.equals(uri.getScheme())) {
            try (Cursor cursor = context.getContentResolver().query(
                    uri,
                    new String[]{OpenableColumns.DISPLAY_NAME},
                    null,
                    null,
                    null)) {
                if (cursor != null && cursor.moveToFirst()) {
                    int index = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
                    if (index >= 0) {
                        String name = cursor.getString(index);
                        if (!TextUtils.isEmpty(name)) {
                            return name;
                        }
                    }
                }
            }
        }
        String lastSegment = uri.getLastPathSegment();
        return TextUtils.isEmpty(lastSegment) ? uri.toString() : lastSegment;
    }

    public static long audioDurationMs(Context context, Uri uri) {
        if (context == null || uri == null) {
            return 0L;
        }
        MediaMetadataRetriever retriever = new MediaMetadataRetriever();
        try {
            retriever.setDataSource(context, uri);
            String duration = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION);
            if (TextUtils.isEmpty(duration)) {
                return 0L;
            }
            return Math.max(0L, Long.parseLong(duration));
        } catch (RuntimeException ignored) {
            return 0L;
        } finally {
            try {
                retriever.release();
            } catch (IOException | RuntimeException ignored) {
            }
        }
    }

    public static String subtitleNameForAudioName(String audioName) {
        return safeName(audioName) + ".vtt";
    }

    public static boolean isSupportedAudioName(String name, String mimeType) {
        String normalized = normalizedName(name);
        if (normalized.endsWith(".vtt")) {
            return false;
        }
        if (mimeType != null && mimeType.startsWith("audio/")) {
            return true;
        }
        return normalized.endsWith(".mp3")
                || normalized.endsWith(".wav")
                || normalized.endsWith(".flac")
                || normalized.endsWith(".m4a")
                || normalized.endsWith(".aac")
                || normalized.endsWith(".ogg")
                || normalized.endsWith(".opus")
                || normalized.endsWith(".amr")
                || normalized.endsWith(".m4b")
                || normalized.endsWith(".3gp");
    }

    private static Intent baseOpenDocumentIntent(String type) {
        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        intent.setType(type);
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION
                | Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION);
        return intent;
    }

    private static List<TreeDocument> treeDocuments(Context context, Uri treeUri) {
        String treeDocumentId = DocumentsContract.getTreeDocumentId(treeUri);
        Uri childrenUri = DocumentsContract.buildChildDocumentsUriUsingTree(treeUri, treeDocumentId);
        List<TreeDocument> documents = new ArrayList<>();
        try (Cursor cursor = context.getContentResolver().query(
                childrenUri,
                FOLDER_COLUMNS,
                null,
                null,
                null)) {
            if (cursor == null) {
                return documents;
            }
            int idIndex = cursor.getColumnIndex(DocumentsContract.Document.COLUMN_DOCUMENT_ID);
            int nameIndex = cursor.getColumnIndex(DocumentsContract.Document.COLUMN_DISPLAY_NAME);
            int mimeIndex = cursor.getColumnIndex(DocumentsContract.Document.COLUMN_MIME_TYPE);
            while (cursor.moveToNext()) {
                String documentId = idIndex >= 0 ? cursor.getString(idIndex) : "";
                String name = nameIndex >= 0 ? cursor.getString(nameIndex) : "";
                String mimeType = mimeIndex >= 0 ? cursor.getString(mimeIndex) : "";
                if (TextUtils.isEmpty(documentId)
                        || TextUtils.isEmpty(name)
                        || DocumentsContract.Document.MIME_TYPE_DIR.equals(mimeType)) {
                    continue;
                }
                documents.add(new TreeDocument(
                        name,
                        mimeType,
                        DocumentsContract.buildDocumentUriUsingTree(treeUri, documentId)));
            }
        } catch (IllegalArgumentException | SecurityException ignored) {
        }
        return documents;
    }

    private static TreeDocument firstMatchingSubtitle(Map<String, TreeDocument> subtitlesByName, String audioName) {
        return subtitlesByName.get(normalizedName(subtitleNameForAudioName(audioName)));
    }

    private static boolean isSupportedSubtitleName(String name) {
        String normalized = normalizedName(name);
        return normalized.endsWith(".vtt");
    }

    private static String normalizedName(String name) {
        return safeName(name).toLowerCase(Locale.ROOT);
    }

    private static String safeName(String name) {
        return name == null ? "" : name.trim();
    }

    public static final class FolderImportItem {
        public final String audioName;
        public final Uri audioUri;
        public final String subtitleName;
        public final Uri subtitleUri;

        public FolderImportItem(String audioName, Uri audioUri, String subtitleName, Uri subtitleUri) {
            this.audioName = audioName == null ? "" : audioName;
            this.audioUri = audioUri;
            this.subtitleName = subtitleName == null ? "" : subtitleName;
            this.subtitleUri = subtitleUri;
        }

        public boolean hasSubtitle() {
            return subtitleUri != null;
        }
    }

    private static final class TreeDocument {
        final String name;
        final String mimeType;
        final Uri uri;

        TreeDocument(String name, String mimeType, Uri uri) {
            this.name = name == null ? "" : name;
            this.mimeType = mimeType == null ? "" : mimeType;
            this.uri = uri;
        }
    }
}

File: app/src/main/java/io/github/summerdez/asmrplayer/data/files/DocumentFiles.kt
package io.github.summerdez.asmrplayer.data.files

import android.content.ClipData
import android.content.ContentResolver
import android.content.Context
import android.content.Intent
import android.database.Cursor
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.provider.DocumentsContract
import android.provider.OpenableColumns
import android.text.TextUtils
import java.io.IOException
import java.util.Collections
import java.util.Locale

object DocumentFiles {
    private val FOLDER_COLUMNS = arrayOf(
        DocumentsContract.Document.COLUMN_DOCUMENT_ID,
        DocumentsContract.Document.COLUMN_DISPLAY_NAME,
        DocumentsContract.Document.COLUMN_MIME_TYPE,
    )

    fun audioPickerIntent(allowMultiple: Boolean): Intent {
        val intent = baseOpenDocumentIntent("audio/*")
        if (allowMultiple) {
            intent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true)
        }
        return intent
    }

    fun folderPickerIntent(): Intent {
        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT_TREE)
        intent.addFlags(
            Intent.FLAG_GRANT_READ_URI_PERMISSION or
                Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION or
                Intent.FLAG_GRANT_PREFIX_URI_PERMISSION,
        )
        return intent
    }

    fun subtitlePickerIntent(): Intent {
        val intent = baseOpenDocumentIntent("*/*")
        intent.putExtra(
            Intent.EXTRA_MIME_TYPES,
            arrayOf(
                "application/x-subrip",
                "text/plain",
                "text/*",
                "application/octet-stream",
            ),
        )
        return intent
    }

    fun imagePickerIntent(): Intent {
        return baseOpenDocumentIntent("image/*")
    }

    fun urisFromResult(data: Intent?): List<Uri> {
        val uris = ArrayList<Uri>()
        val clipData: ClipData? = data!!.clipData
        if (clipData != null) {
            for (i in 0 until clipData.itemCount) {
                val uri = clipData.getItemAt(i).uri
                if (uri != null) {
                    uris.add(uri)
                }
            }
        } else if (data.data != null) {
            uris.add(data.data!!)
        }
        return uris
    }

    fun persistReadPermission(context: Context?, uri: Uri?) {
        try {
            context!!.contentResolver.takePersistableUriPermission(
                platformValue(uri),
                Intent.FLAG_GRANT_READ_URI_PERMISSION,
            )
        } catch (ignored: SecurityException) {
        }
    }

    fun persistTreeReadPermission(context: Context?, data: Intent?, uri: Uri?) {
        val flags = data?.flags ?: Intent.FLAG_GRANT_READ_URI_PERMISSION
        var readFlags = flags and Intent.FLAG_GRANT_READ_URI_PERMISSION
        if (readFlags == 0) {
            readFlags = Intent.FLAG_GRANT_READ_URI_PERMISSION
        }
        try {
            context!!.contentResolver.takePersistableUriPermission(platformValue(uri), readFlags)
        } catch (ignored: SecurityException) {
        }
    }

    fun folderAudioImports(context: Context?, treeUri: Uri?): List<FolderImportItem> {
        if (treeUri == null) {
            return Collections.emptyList()
        }
        val documents = treeDocuments(context, treeUri)
        val subtitlesByName = HashMap<String, TreeDocument>()
        for (document in documents) {
            if (isSupportedSubtitleName(document.name)) {
                subtitlesByName[normalizedName(document.name)] = document
            }
        }

        val imports = ArrayList<FolderImportItem>()
        for (document in documents) {
            if (!isSupportedAudioName(document.name, document.mimeType)) {
                continue
            }
            val subtitle = firstMatchingSubtitle(subtitlesByName, document.name)
            imports.add(
                FolderImportItem(
                    document.name,
                    document.uri,
                    subtitle?.name ?: "",
                    subtitle?.uri,
                ),
            )
        }
        imports.sortWith { left, right -> String.CASE_INSENSITIVE_ORDER.compare(left.audioName, right.audioName) }
        return imports
    }

    fun displayName(context: Context?, uri: Uri?): String {
        val targetUri = uri!!
        if (ContentResolver.SCHEME_CONTENT == targetUri.scheme) {
            context!!.contentResolver.query(
                targetUri,
                arrayOf(OpenableColumns.DISPLAY_NAME),
                null,
                null,
                null,
            )?.use { cursor ->
                if (cursor.moveToFirst()) {
                    val index = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                    if (index >= 0) {
                        val name = cursor.getString(index)
                        if (!TextUtils.isEmpty(name)) {
                            return name
                        }
                    }
                }
            }
        }
        val lastSegment = targetUri.lastPathSegment
        return if (TextUtils.isEmpty(lastSegment)) targetUri.toString() else lastSegment!!
    }

    fun audioDurationMs(context: Context?, uri: Uri?): Long {
        if (context == null || uri == null) {
            return 0L
        }
        val retriever = MediaMetadataRetriever()
        try {
            retriever.setDataSource(context, uri)
            val duration = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)
            if (TextUtils.isEmpty(duration)) {
                return 0L
            }
            return maxOf(0L, java.lang.Long.parseLong(duration!!))
        } catch (ignored: RuntimeException) {
            return 0L
        } finally {
            try {
                retriever.release()
            } catch (ignored: IOException) {
            } catch (ignored: RuntimeException) {
            }
        }
    }

    fun subtitleNameForAudioName(audioName: String?): String {
        return safeName(audioName) + ".vtt"
    }

    fun isSupportedAudioName(name: String?, mimeType: String?): Boolean {
        val normalized = normalizedName(name)
        if (normalized.endsWith(".vtt")) {
            return false
        }
        if (mimeType != null && mimeType.startsWith("audio/")) {
            return true
        }
        return normalized.endsWith(".mp3") ||
            normalized.endsWith(".wav") ||
            normalized.endsWith(".flac") ||
            normalized.endsWith(".m4a") ||
            normalized.endsWith(".aac") ||
            normalized.endsWith(".ogg") ||
            normalized.endsWith(".opus") ||
            normalized.endsWith(".amr") ||
            normalized.endsWith(".m4b") ||
            normalized.endsWith(".3gp")
    }

    private fun baseOpenDocumentIntent(type: String): Intent {
        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
        intent.addCategory(Intent.CATEGORY_OPENABLE)
        intent.type = type
        intent.addFlags(
            Intent.FLAG_GRANT_READ_URI_PERMISSION or
                Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION,
        )
        return intent
    }

    private fun treeDocuments(context: Context?, treeUri: Uri): List<TreeDocument> {
        val treeDocumentId = DocumentsContract.getTreeDocumentId(treeUri)
        val childrenUri = DocumentsContract.buildChildDocumentsUriUsingTree(treeUri, treeDocumentId)
        val documents = ArrayList<TreeDocument>()
        try {
            context!!.contentResolver.query(
                childrenUri,
                FOLDER_COLUMNS,
                null,
                null,
                null,
            )?.use { cursor ->
                collectTreeDocuments(cursor, treeUri, documents)
            } ?: return documents
        } catch (ignored: IllegalArgumentException) {
        } catch (ignored: SecurityException) {
        }
        return documents
    }

    private fun collectTreeDocuments(cursor: Cursor, treeUri: Uri, documents: MutableList<TreeDocument>) {
        val idIndex = cursor.getColumnIndex(DocumentsContract.Document.COLUMN_DOCUMENT_ID)
        val nameIndex = cursor.getColumnIndex(DocumentsContract.Document.COLUMN_DISPLAY_NAME)
        val mimeIndex = cursor.getColumnIndex(DocumentsContract.Document.COLUMN_MIME_TYPE)
        while (cursor.moveToNext()) {
            val documentId = if (idIndex >= 0) cursor.getString(idIndex) else ""
            val name = if (nameIndex >= 0) cursor.getString(nameIndex) else ""
            val mimeType = if (mimeIndex >= 0) cursor.getString(mimeIndex) else ""
            if (
                TextUtils.isEmpty(documentId) ||
                TextUtils.isEmpty(name) ||
                DocumentsContract.Document.MIME_TYPE_DIR == mimeType
            ) {
                continue
            }
            documents.add(
                TreeDocument(
                    name,
                    mimeType,
                    DocumentsContract.buildDocumentUriUsingTree(treeUri, documentId),
                ),
            )
        }
    }

    private fun firstMatchingSubtitle(
        subtitlesByName: Map<String, TreeDocument>,
        audioName: String?,
    ): TreeDocument? {
        return subtitlesByName[normalizedName(subtitleNameForAudioName(audioName))]
    }

    private fun isSupportedSubtitleName(name: String?): Boolean {
        val normalized = normalizedName(name)
        return normalized.endsWith(".vtt")
    }

    private fun normalizedName(name: String?): String {
        return safeName(name).lowercase(Locale.ROOT)
    }

    private fun safeName(name: String?): String {
        return name?.trim() ?: ""
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T> platformValue(value: T?): T {
        return value as T
    }

    class FolderImportItem(
        audioName: String?,
        audioUri: Uri?,
        subtitleName: String?,
        val subtitleUri: Uri?,
    ) {
        val audioName: String = audioName ?: ""

        val audioUri: Uri = run {
            @Suppress("UNCHECKED_CAST")
            fun <T> platformFieldValue(value: T?): T {
                return value as T
            }

            platformFieldValue(audioUri)
        }

        val subtitleName: String = subtitleName ?: ""

        fun hasSubtitle(): Boolean {
            return subtitleUri != null
        }
    }

    private class TreeDocument(
        name: String?,
        mimeType: String?,
        val uri: Uri?,
    ) {
        val name: String = name ?: ""
        val mimeType: String = mimeType ?: ""
    }
}
